### ISSUE TITLE
More fine-grain selection of assembly code paths

### ISSUE THREAD

**Issue by GrandAdmiralThrawn** [OpenVisualCloud/SVT-HEVC#262]
While I'm aware that this is a high performance library focused on AVX2/AVX512, I would like to pick assembly code paths more selectively, or to have a better auto-detection.

Currently, SVT-HEVC allows the user to pick AVX512 (-asm 7), AVX2 (-asm 3), NOAVX2 (-asm 1) and pure C code (-asm 0). However, "NOAVX2" still doesn't make the code run on machines without any AVX (or maybe BMI? FMA?). So there is no solution in between pure C and AVX2, even though lots of SSE code seems to be present as well.

The last release version code can run in pure C mode on much older CPUs as well after all, it's just much, much slower.

What I would like to have is the ability to run it purely with SSE2 optimizations on the first x86_64 generations, or purely with SSSE3 or SSE4.1 optimizations on the later ones, depending on the CPU at hand. Similarly, one could also choose whether to use or to leave out AVX2 or AVX512 where it's present, similar to what's possible with the x265 [--asm](https://x265.readthedocs.io/en/default/cli.html#cmdoption-asm) command line switch. That would also allow for easy performance comparisons between instruction set extensions.

I'd be fine with the code path choice happening at compile time too. I would just like to be able to make better use of CPUs without any AVX as well, falling back to SSE2/SSSE3/SSE4.1 code where necessary.

PS.: A compile-time cmake knob to en-/disable NUMA support would also be nice for older operating systems. I found the necessary change in the source code to be absolutely trivial, after which it can run on non-NUMA aware systems as well.

Thanks for your work!

**Comment by tianjunwork** [comment]
In terms of your request of fine-grain selection of SIMD architectures to use and falling back to SSEx, SVT team has no plan to support it in the near future. Contribution from community is highly appreciated.
Could you tell me more on en/disable NUMA in cmake?

**Comment by Austin-Hu** [comment]
Hi @GrandAdmiralThrawn ,

At first, currently the "-asm" option of SVT-HEVC only supports 0 or 1 ([0: C Only, 1: Auto]), according to the comments in Source/API/EbApi.h: "1 = up to AVX512, auto-select highest assembly instruction set supported.".

And there would be much effort to implement the key "kernel" functions with all instruction sets (AVX512, AVX2, SSEx). In addition to the C_DEFAULT implementation, most functions only use one kind of instruction set, for example, IntraAngVertical_funcPtrArray only has AVX2, meanwhile AvcStyleUniPredLumaIFFunctionPtrArray only has SSEx implementation. BTW, you can search for all of those functions with ASM_TYPES.

So it's not in our plan yet.

**Comment by GrandAdmiralThrawn** [comment]
I thought that you would think along that way, given than older and deprecated architectures are likely not the focus of SVT-HEVC. In the meantime I managed to re-enable some SSE2 code, but the speed gains were below 10% so far.. and my hacks are extremely dirty and bad.

About NUMA: I didn't do this via cmake, I hacked up your code directly, in a very, very ugly and bad (and likely also redundant) way.

**App/EbAppMain.c**, function `AssignAppThreadGroup()`, remove the following lines to not call NUMA-aware functions:
```
    if (GetActiveProcessorGroupCount() == 2) {
        GROUP_AFFINITY           groupAffinity;
        GetThreadGroupAffinity(GetCurrentThread(), &groupAffinity);
        groupAffinity.Group = targetSocket;
        SetThreadGroupAffinity(GetCurrentThread(), &groupAffinity, NULL);
    }
```


**Lib/Codec/EbEncHandle.c**, function `InitThreadManagmentParams()`, remove NUMA functions:
```
    GetThreadGroupAffinity(GetCurrentThread(),&groupAffinity);
    numGroups = (EB_U8) GetActiveProcessorGroupCount();

```
Set the number of processor groups to 1 statically afterwards:
```
	numGroups = (EB_U8) 1;
```


**Lib/Codec/EbThreads.c**, `#define EB_CREATETHREAD()`, remove NUMA group functionality and just replace it with regular, flat 64b bit-vector cpu addressing:

Remove:
```
        if(numGroups == 1) {\
            SetThreadAffinityMask(pointer, groupAffinity.Mask);\
        }\
        else if (numGroups == 2 && alternateGroups){ \
            groupAffinity.Group = 1 - groupAffinity.Group; \
            SetThreadGroupAffinity(pointer,&groupAffinity,NULL); \
        } \
        else if (numGroups == 2 && !alternateGroups){ \
            SetThreadGroupAffinity(pointer,&groupAffinity,NULL); \
        } \
```
Replace with:
```
        SetThreadAffinityMask(pointer, groupAffinity.Mask);\
```

With those changes (and a lot more for disabling AVX stuff) I have been able to back-port SVT-HEVC successfully to systems without NUMA and without OS XSAVE/XRSTOR instructions (Vista x64 and even XP x64).

But the changes are extremely bad - running my version on a modern system doesn't allow the user to use NUMA anymore, and I guess SSSE3, SSE4.1, AVX2 & AVX-512 would also no longer work.

I'm not a C programmer or even a programmer in general at all (I don't understand even half of your code), so it's like a gardener trying to re-assemble a modern turbine engine to upgrade his old lawnmower from the 60's. ;)

**Comment by GrandAdmiralThrawn** [comment]
On 2019/07/26 03:13, Austin-Hu wrote:
> Hi @GrandAdmiralThrawn <https://github.com/GrandAdmiralThrawn> ,
> 
> At first, currently the "-asm" option of SVT-HEVC only supports 0 or 1 ([0: C Only, 1:
> Auto]), according to the comments in Source/API/EbApi.h: "1 = up to AVX512, auto-select
> highest assembly instruction set supported.".
> 
> And there would be much effort to implement the key "kernel" functions with all
> instruction sets (AVX512, AVX2, SSEx). In addition to the C_DEFAULT implementation, most
> functions only use one kind of instruction set, for example, IntraAngVertical_funcPtrArray
> only has AVX2, meanwhile AvcStyleUniPredLumaIFFunctionPtrArray only has SSEx
> implementation. BTW, you can search for all of those functions with ASM_TYPES.
Ah yes, you're right regarding `--asm`! The bit mask is in the code, but the command line
parameter only supports pure C or auto-detection. My mistake, I just assumed something
that wasn't there.

Also, I am not asking for a full implemention of all key functions in all instruction
sets. I mean, that'd be awesome, but it's ok if it's not done. I would just like to tell
the build system (or the program/lib at runtime): Here you have a SSE4.1 CPU, do what you
can as fast as you can on that chip (With the few SSE4.1 / SSSE3 / SSE2 code paths that
are already there).

Even if only small parts are run with SSEx then, it'd still be better than forcing users
with older CPUs to use pure C (which is what the code currently seems to be doing).

Thanks!

**Comment by GrandAdmiralThrawn** [comment]
Alright, one more. I managed to re-enable SSSE3 and SSE4.1 code and leave out just the AVX2 and AVX512 parts. Essentially I made the CPU detection code pretend I have AVX2/AVX512 (`EB_U32 asmType = 7;`), and then I just replaced all calls of AVX2 & AVX512 functions with SSEx ones where available and with pure C ones where necessary - a terrible way to do this, I know.

CPU was a Core i7 980X 3.33GHz with HT disabled (because VirtualBox VM), Guest OS was XP x64 SP2. RAM was plenty - 8GiB.

For a 1280×720p sample (2878 frames), going from pure C to SSE2-only improved speed by roughly +7.5%. Going from pure C to SSE2+SSSE3+SSE4.1 improved speed by roughly +13%. That's really not much of a gain I guess, but it's at least something.

I will look into doing a proper code path for this in SVT-HEVC, but since I suck so much at C, I might fail to do this cleanly...

**Comment by Austin-Hu** [comment]
Hi @GrandAdmiralThrawn ,

For some CPU (such as the Core i7 980X you have) which couldn't support AVX512 or AVX2, the SSEx version has be implemented, for the functions with such filtering "!!(ASM_TYPES & AVX2_MASK)" or "!!(ASM_TYPES & AVX512_MASK)", in order to take advantage of the capability of the specific CPU, rather than just falling back to C_DEFAULT implementation, as you're doing right now.

I think the ideal way to make use of the instruction sets for any dedicated CPU, is to implement all versions (AVX512, AVX2, SSEx, etc.) for the key functions, and to assign ASM_TYPES with any detected type in EbInitEncoder(), rather than the bit map. So the supported version for those functions would be used during run time later. But you know it would take much effort.

**Comment by GrandAdmiralThrawn** [comment]
On 2019/07/29 03:28, Austin-Hu wrote:> Hi @GrandAdmiralThrawn ,
>
> For some CPU (such as the Core i7 980X you have) which couldn't support AVX512 or AVX2,
the SSEx version has be implemented, for the functions with such filtering "!!(ASM_TYPES &
AVX2_MASK)" or "!!(ASM_TYPES & AVX512_MASK)", in order to take advantage of the capability
of the specific CPU, rather than just falling back to C_DEFAULT implementation, as you're
doing right now.
>
> I think the ideal way to make use of the instruction sets for any dedicated CPU, is to
implement all versions (AVX512, AVX2, SSEx, etc.) for the key functions, and to assign
ASM_TYPES with any detected type in EbInitEncoder(), rather than the bit map. So the
supported version for those functions would be used during run time later. But you know it
would take much effort.

Hello @Austin-Hu

Well, before starting to mess around, I just compiled the code and tried to run it on said
i7 980X, and it crashed hard. Debugged/disassembled it, and it clearly showed AVX2
instructions being called.

After inspecting the code I also found that sometimes, SSSE3 calls would be placed in the
AVX2 code path, but not in the NONAVX2 one. So meanwhile, i managed to remove all
AVX2/AVX512 code and re-enable SSE2+SSSE3+SSE4.1 for my own machines. My mess is clearly
not redistributable in its current state, but at least it works for me - which the default
SVT-HEVC didn't.

It's not much faster (roughly +13% in comparison to pure C), but better than nothing. ;)

**Comment by Austin-Hu** [comment]
Hi @GrandAdmiralThrawn , good to know that. Could you share your patch or private repository once you think it's OK, so that we can also know where the incorrect AVX2 usages are, and watch your progress? Thanks!

**Comment by tianjunwork** [comment]
Hi @GrandAdmiralThrawn, about your changes of the thread affinity, numGroups is assigned by below function. Based on value of numGroups, thready affinity is set. It should serve your purpose.
```numGroups = (EB_U8) GetActiveProcessorGroupCount();```
https://docs.microsoft.com/en-us/windows/win32/procthread/processor-groups
Systems with fewer than 64 logical processors always have a single group, Group 0.

**Comment by GrandAdmiralThrawn** [comment]
@tianjunwork Thank you! I've already found that part and blatantly set numGroups to 1 in EbEncHandle.c. The EB_CREATETHREAD macro in EbThreads.h also causes problems however, when the code is run on a non-NUMA-aware OS. Despite the if clauses in there, it still tries to call `SetThreadGroupAffinity()`, even though I don't get why. But then again, I know nothing about how the compiler actually treats macros... ;)

Also in EbEncHandle.c, there is a function `InitThreadManagmentParams()`. That function just calls `GetThreadGroupAffinity()` on all Windows systems without checking for the presence of NUMA. Since that function does not exist on Windows systems without NUMA, the code fails.

@Austin-Hu I'm uncomfortable with sharing my trash modifications, as I've pretty much messed up your code. I don't even know how to properly create .diff patch files, and I don't have a repository. As I said, I'm not a developer and I don't really know how to program, I'm just a user.

What I could do is to re-compile the stock version, and run it through the debugger, noting down all occurrences of AVX2 code being run unconditionally, and then post a quick list of affected functions / code parts.

**Comment by tianjunwork** [comment]
@GrandAdmiralThrawn, the encoder runs fine on Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz which with coreinfo.exe shows:
Logical Processor to Socket Map:
****  Socket 0
Logical Processor to NUMA Node Map:
****  NUMA Node 0
**No NUMA nodes.**
Logical Processor to Group Map:
****  Group 0
SetThreadGroupAffinity() exists on all windows release and works on No NUMA system.

**Comment by GrandAdmiralThrawn** [comment]
@tianjunwork Ah, maybe I didn't express myself properly when it comes to the NUMA stuff.

Yes, `SetThreadGroupAffinity()` exists on all no-NUMA hardware platforms as long as a NUMA-aware operating system is installed. So, in the Microsoft case, Windows 7, 8.x and 10. But that function does not exist on operating systems that know nothing about NUMA - like older Vista x64 or XP x64. See [[here](https://docs.microsoft.com/en-us/windows/win32/api/processtopologyapi/nf-processtopologyapi-setthreadgroupaffinity)].

Of course, as I said above, those are no longer supported by Microsoft, so I cannot expect you to support them either. It would just be a "nice to have" and it's probably easy to do in ways that aren't as ugly as mine.

(Yes, I love to make modern code run on old platforms.)

**Comment by tianjunwork** [comment]
@GrandAdmiralThrawn Thx for the information you provided:) And thank you for the understanding of our priority. We have no plan of supporting Vista x64 or XP x64.

**Comment by GrandAdmiralThrawn** [comment]
@tianjunwork Thought so. ;) Well, then I'll just keep backporting it by myself, the NUMA part is super easy after all. Thanks!
